proposals/NNNN-noncopyable-generics.md

```diff
@@ -0,0 +1,713 @@
+
+```swift
+enum List<Element /* : Copyable */> /* : Copyable */ {
+  case empty
+  indirect case node(Element, List<Element>)
+}
+```
+
+As in SE-390, a generic struct or enum can use `~Copyable` to suppress the implicit `Copyable` conformance and constraint:
+
+```swift
+enum List<Elm: ~Copyable>: ~Copyable { /* ... */ }
+
+// or equivalently:
+
+enum List<Elm>: ~Copyable where Elm: ~Copyable { /* ... */ }
+```
+
+Since `Elm` is not required to be `Copyable`, a noncopyable type like `FileDescriptor` can be substituted in `List`, in addition to copyable ones.
+
+### Classes
+
+Classes (including actors) are always `Copyable`,
+even if they have noncopyable properties.
+This proposal does not provide support for noncopyable classes.
+```swift
+class FileHandle<File> /* : Copyable where File: Copyable */ {
+  var file: File
+  // ...
+}
+```
+
+Generic parameters on classes are `Copyable` by default.
+This can be suppressed with `~Copyable` on the type parameter:
+```swift
+class FileHandle<File: ~Copyable> /* : Copyable */ {
+  var file: File
+  // ...
+}
+```
+
+### Protocols
+
+Protocols and their associated types default to implicit `Copyable` conformance requirement:
+
+```swift
+// signature <Self where Self: Copyable, Self.T: Copyable>
+protocol Foo /* : Copyable */ {
+  associatedtype T /* : Copyable */
+
+  borrowing func bar() -> Self
+  func buzz(_: T) -> T
+  func blarg() -> RequiresCopyable<Self>
+}
+```
+
+Protocols can suppress the default `Copyable` requirement from `Self` using `~Copyable`:
+
+```swift
+// signature <Self where Self.Event: Copyable>
+protocol EventLog: ~Copyable {
+  associatedtype Event /* : Copyable */
+
+  mutating func push(_ event: Event)
+  mutating func pop() throws -> Event
+}
+```
+
+Within `EventLog`, the type `Self` has no conformance requirements at all, but the associated type `Self.Event` is copyable.
+Suppressing the `Copyable` conformance requirement on `EventLog` allows copyable and noncopyable types to conform:
+
+```swift
+// signature <Self where Self: EventLog, Self: Copyable>
+struct ArrayLog<Element>: EventLog /*, Copyable where Element: Copyable */ {
+  typealias Event = Element
+  var log: [Element]
+  // ...
+}
+
+// signature <Self where Self: EventLog>
+struct UniqueLog<Element>: EventLog, ~Copyable /* where Element: Copyable */ {
+  typealias Event = Element
+  var log: [Element]
+  // ...
+}
+```
+
+Associated types can additionally use `~Copyable` to suppress their default `Copyable` requirement, allowing a noncopyable type to witness the requirement:
+
+```swift
+protocol JobQueue<Job> /* : Copyable */ {
+  associatedtype Job: ~Copyable
+
+  func submit(_ job: consuming Job)
+}
+```
+
+#### Protocol Extensions
+
+An extension of a protocol is implicitly constrained to `Self: Copyable` unless you explicitly state otherwise.
+```swift
+protocol EventLog: ~Copyable {
+  ...
+}
+
+extension EventLog /* where Self: Copyable */ {
```

Comment:
Is `Copyable` also the default in protocol extensions for associated types that are `~Copyable`? Can you please specify the rule and add an example?